sno/gpkg_adapter.py

```diff
@@ -0,0 +1,251 @@
+
+
+def generate_gpkg_contents(dataset):
+    """Generate a gpkg_contents meta item from a dataset."""
+    is_spatial = bool(_get_geometry_columns(dataset.schema))
+
+    result = {
+        "identifier": dataset.get_meta_item("title"),
+        "description": dataset.get_meta_item("description"),
+        "table_name": dataset.tree.name,
+        "data_type": "features" if is_spatial else "attributes",
+    }
+    if is_spatial:
+        result["srs_id"] = _gpkg_srs_id(dataset)
+    return result
+
+
+def generate_gpkg_geometry_columns(dataset):
+    """Generate a gpkg_geometry_columns meta item from a dataset."""
+    geom_columns = _get_geometry_columns(dataset.schema)
+    if not geom_columns:
+        return None
+
+    geometry_type = geom_columns[0].extra_type_info["geometryType"]
+    type_name, *zm = geometry_type.split(" ", 1)
+    zm = zm[0] if zm else ""
+    z = 1 if "Z" in zm else 0
+    m = 1 if "M" in zm else 0
+
+    return {
+        "table_name": dataset.tree.name,
+        "column_name": geom_columns[0].name,
+        "geometry_type_name": type_name,
+        "srs_id": _gpkg_srs_id(dataset),
+        "z": z,
+        "m": m,
+    }
+
+
+def generate_gpkg_spatial_ref_sys(dataset):
+    """Generate a gpkg_spatial_ref_sys meta item from a dataset."""
+    geom_columns = _get_geometry_columns(dataset.schema)
+    if not geom_columns:
+        return []
+
+    srs_pathname = geom_columns[0].extra_type_info["geometrySRS"]
+    if not srs_pathname:
+        return []
+    definition = dataset.get_srs_definition(srs_pathname)
+    return wkt_to_gpkg_spatial_ref_sys(definition)
+
+
+def _gpkg_srs_id(dataset):
+    gsrs = dataset.get_meta_item("gpkg_spatial_ref_sys")
+    return gsrs[0]["srs_id"] if gsrs else 0
+
+
+def wkt_to_gpkg_spatial_ref_sys(wkt):
+    """Given a WKT srs definition, generate a gpkg_spatial_ref_sys meta item."""
+    return _gpkg_spatial_ref_sys(SpatialReference(wkt), wkt)
+
+
+def osgeo_to_gpkg_spatial_ref_sys(spatial_ref):
+    """Given an osgeo SpatialReference, generate a gpkg_spatial_ref_sys meta item."""
+    return _gpkg_spatial_ref_sys(spatial_ref, spatial_ref.ExportToWkt())
+
+
+def _gpkg_spatial_ref_sys(spatial_ref, wkt):
+    # TODO: Better support for custom WKT. https://github.com/koordinates/sno/issues/148
+    spatial_ref.AutoIdentifyEPSG()
+    organization = spatial_ref.GetAuthorityName(None) or "NONE"
+    srs_id = spatial_ref.GetAuthorityCode(None) or 0
+    return [
+        {
+            "srs_name": spatial_ref.GetName(),
+            "definition": wkt,
+            "organization": organization,
+            "srs_id": srs_id,
+            "organization_coordsys_id": srs_id,
+            "description": None,
+        }
+    ]
+
+
+def wkt_to_srs_str(wkt):
+    """Given a WKT srs definition, generate a sensible name for it."""
+    return osgeo_to_srs_str(SpatialReference(wkt))
+
+
+def osgeo_to_srs_str(spatial_ref):
+    """Given a osgeo SpatialReference, generate a sensible name for it."""
```

Comment:
```suggestion
    """Given a osgeo SpatialReference, generate a sensible identifier for it."""
```

FYI, a human-readable name/description comes from the top-level PROJCS/GEOGCS element, and it's available via `srs.GetName()`.

```python
>>> srs = osr.SpatialReference()
>>> srs.ImportFromEPSG(2193)
>>> srs.GetName()
'NZGD2000 / New Zealand Transverse Mercator 2000'
```

Reply:
Done